REGRESSION (285996@main): Readjusting the text selection boundaries moves the highlighted text box on Scribd.com

rniwa opened the pull request:
#### 668204deb7e9719144e14d5ad2a8612c20d7575a
<pre>
REGRESSION (285996@main): Readjusting the text selection boundaries moves the highlighted text box on Scribd.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=286590">https://bugs.webkit.org/show_bug.cgi?id=286590</a>

Reviewed by Megan Gardner and Wenson Hsieh.

Add a quirk for scribd.com to preserve the behavior prior to 285996@main.

The crux of this issue is that the selection extent momentarily moves to the beginning of the document when the user
extends the selection downwards. This causes website&apos;s code to reverse selection base and extent to kick in.
The same bug exists in macOS Safari as well as Firefox when extending upwards / backwards.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::setSelectionWithoutUpdatingAppearance):
(WebCore::FrameSelection::updateOrDisassociateLiveRange):
* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldReuseLiveRangeForSelectionUpdate const):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:

Canonical link: <a href="https://commits.webkit.org/289483@main">https://commits.webkit.org/289483@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/16f05839c50bed73098a9d0705409aa4c4b606bd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87034 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6544 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41389 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91892 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37773 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89083 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6820 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14612 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67273 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25032 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90036 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5219 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78780 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47595 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4996 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33155 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36890 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75488 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34032 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93780 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14196 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10331 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76077 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14400 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74633 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75277 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19613 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18041 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7113 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13565 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14215 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19508 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13959 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17402 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15740 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1735,6 +1735,11 @@ bool Quirks::shouldAvoidStartingSelectionOnMouseDown(const Node& target) const
 
+bool Quirks::shouldReuseLiveRangeForSelectionUpdate() const
+{
+    return needsQuirks() && isDomain("scribd.com"_s);
```
Perhaps it's worth adding a cached bit for this on `QuirksData`, like some of the other quirks? (Since the codepath that calls into this can be invoked quite frequently)

rniwa replied:
Yeah, we can do that. I was thinking of doing it anyway.

rniwa committed:
REGRESSION (285996@main): Readjusting the text selection boundaries moves the highlighted text box on Scribd.com
https://bugs.webkit.org/show_bug.cgi?id=286590

Reviewed by Megan Gardner and Wenson Hsieh.

Add a quirk for scribd.com to preserve the behavior prior to 285996@main.

The crux of this issue is that the selection extent momentarily moves to the beginning of the document when the user
extends the selection downwards. This causes website's code to reverse selection base and extent to kick in.
The same bug exists in macOS Safari as well as Firefox when extending upwards / backwards.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::setSelectionWithoutUpdatingAppearance):
(WebCore::FrameSelection::updateOrDisassociateLiveRange):
* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldReuseLiveRangeForSelectionUpdate const):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:

Canonical link: https://commits.webkit.org/289483@main
